packages/core/useWebLocks/index.ts

```diff
@@ -0,0 +1,113 @@
+   * @return Promise that resolves when the lock is released or gets rejected when there was an error getting the lock.
+   */
+  request: ((name: Keys, callback: (signal: AbortSignal) => Promise<unknown> | unknown) => Promise<unknown>) & ((name: Keys, options: UseWebLocksRequestOptions, callback: (signal: AbortSignal) => Promise<unknown> | unknown) => Promise<unknown>)
+}
+
+/**
+ * Reactive Web Locks API. The lock automatically gets release on scope dispose.
+ *
+ * @see https://vueuse.org/useWebLocks/
+ * @param [options]
+ */
+export function useWebLocks<Keys extends string = string>(options: UseWebLocksOptions = {}): UseWebLocksReturn<Keys> {
+  const { navigator = defaultNavigator } = options
+  const isSupported = useSupported(() => navigator?.locks && typeof AbortController === 'function')
+  const releaseLocks = new Map<Keys, (value: unknown) => void>()
+  const abortControllers = new Set<AbortController>()
+  function request(name: Keys, callback: (signal: AbortSignal) => Promise<unknown> | unknown): Promise<unknown>
+  function request(name: Keys, options: UseWebLocksRequestOptions, callback: (signal: AbortSignal) => Promise<unknown> | unknown): Promise<unknown>
+  function request(name: Keys, optionsOrFn: UseWebLocksRequestOptions | ((signal: AbortSignal) => Promise<unknown> | unknown), callback?: (signal: AbortSignal) => Promise<unknown> | unknown): Promise<unknown> {
+    if (!isSupported.value) {
+      return Promise.reject(new Error('Web Locks API or AbortController not supported'))
+    }
+    const requestOptions = typeof optionsOrFn === 'function' ? {} : optionsOrFn
+    callback = typeof optionsOrFn === 'function' ? optionsOrFn : callback
+    if (typeof callback !== 'function') {
+      return Promise.reject(new Error('callback not provided'))
+    }
+    const { mode = 'exclusive', ifAvailable = false, steal = false } = requestOptions
+    const opts: LockOptions = { mode, ifAvailable, steal }
+    const abortController = new AbortController()
+    abortControllers.add(abortController)
+    if (!ifAvailable) {
+      opts.signal = abortController.signal
+    }
+    return navigator!.locks.request(name, opts, (lock) => {
+      if (!lock) {
+        if (requestOptions?.onUnavailable) {
+          requestOptions.onUnavailable()
+        }
+        return
+      }
+      return new Promise((resolve, reject) => {
+        releaseLocks.set(name, resolve)
+        Promise.resolve(callback(abortController.signal)).then((onfullfilled) => {
+          releaseLocks.delete(name)
+          abortControllers.delete(abortController)
+          resolve(onfullfilled)
+        }, (onrejected) => {
+          releaseLocks.delete(name)
+          abortControllers.delete(abortController)
+          reject(onrejected)
+        })
+      })
+    })
+  }
+  tryOnScopeDispose(() => {
+    abortControllers.forEach(abortController => abortController.abort())
+    abortControllers.clear()
```

Comment:
Question: Is there a scenario where we only want to abort one controller? otherwise it seems we can use one abort controller for all lock requests. So we do not need to loop.
 
EDIT: Ok, maybe if we have several different reqeusts. But I still think we can use something like
 `AbortSignal.any` here to have an internal signal for all that we abort on dispose and we can create one per request to abort only one when needed.
 
 Ref: https://kettanaito.com/blog/dont-sleep-on-abort-controller

Reply:
I switched to one shared `AbortController` – and added documentation/demo that makes clear that the `request` callback is responsible to clean up event handlers to the `abort` event (when it adds one).

Reply:
We just passing the signal. Not the whole controller. okay. so this should be fine.

Reply:
[AbortSignal.any](https://developer.mozilla.org/de/docs/Web/API/AbortSignal/any_static) is kinda new (baseline 2024). Should we require it in the `isSupported` check or use a [ponyfill](https://gist.github.com/CNSeniorious000/9fc1a72e45358dd7c9e2f16e5d26df5c)?

Reply:
> [AbortSignal.any](https://developer.mozilla.org/de/docs/Web/API/AbortSignal/any_static) is kinda new (baseline 2024). Should we require it in the `isSupported` check or use a [ponyfill](https://gist.github.com/CNSeniorious000/9fc1a72e45358dd7c9e2f16e5d26df5c)?

I think we dont need this, because we will currently dont have an external signal. I have removed my prev comments :)